src/dotnet-scaffolding/Microsoft.DotNet.Scaffolding.Internal/Services/TelemetryService.cs

```diff
@@ -0,0 +1,149 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.Channel;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.DotNet.PlatformAbstractions;
+using Microsoft.DotNet.Scaffolding.Internal.Telemetry;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.Scaffolding.Internal.Services;
+
+internal class TelemetryService : ITelemetryService, IDisposable
+{
+    private readonly IFirstTimeUseNoticeSentinel _firstTimeUseNoticeSentinel;
+    private readonly IEnvironmentService _environmentService;
+    private readonly ILogger _logger;
+    private readonly string? _currentSessionId;
+    private TelemetryClient? _client;
+    private TelemetryConfiguration? _telemetryConfig;
+    private Dictionary<string, string> _commonProperties = [];
+    private Dictionary<string, double> _commonMeasurements = [];
+    private bool _enabled;
+    private Task? _trackEventTask;
+
+    public TelemetryService(IFirstTimeUseNoticeSentinel firstTimeUseNoticeSentinel, IEnvironmentService environmentService, ILogger<TelemetryService> logger)
+    {
+        _firstTimeUseNoticeSentinel = firstTimeUseNoticeSentinel;
+        _environmentService = environmentService;
+        _logger = logger;
+        var dotnetScaffoldTelemetryState = _environmentService.GetEnvironmentVariable(TelemetryConstants.DOTNET_SCAFFOLD_TELEMETRY_STATE);
+        var dotnetScaffoldTelemetryEnabled = !string.IsNullOrEmpty(dotnetScaffoldTelemetryState) && dotnetScaffoldTelemetryState.Equals(TelemetryConstants.TELEMETRY_STATE_ENABLED, StringComparison.OrdinalIgnoreCase);
+
+        _enabled = !_environmentService.GetEnvironmentVariableAsBool(TelemetryConstants.TELEMETRY_OPTOUT) &&
+           (_firstTimeUseNoticeSentinel.Exists() || dotnetScaffoldTelemetryEnabled);
+        if (_enabled)
+        {
+            // Store the session ID in a static field so that it can be reused
+            _currentSessionId = Guid.NewGuid().ToString();
+            _trackEventTask = Task.Factory.StartNew(() => InitializeClient(), CancellationToken.None, TaskCreationOptions.None, TaskScheduler.Default);
+        }
+    }
+
+    public void TrackEvent(string eventName, IReadOnlyDictionary<string, string> properties, IReadOnlyDictionary<string, double> measurements)
+    {
+        if (_trackEventTask is not null)
+        {
+            _trackEventTask = _trackEventTask.ContinueWith(
+                x => TrackEventTask(eventName, properties, measurements),
+                TaskScheduler.Default
+            );
+
+            //wait on the task to finish
+            Flush();
```

Comment:
We don't want to wait on the task here - that would negate using the task because it would block every call to TrackEvent. We would need to expose it via the interface and call it from somewhere after all the telemetry had been sent. In my testing for just dotnet-scaffold itself I wired it into ScaffoldCommand.ExecuteAsync before returning the result. But there is probably a better location (and possibly one that is more consistent between the tools since that won't exist in the other two).

Reply:
ah i see, will do

Reply:
added `IFirstPartyToolTelemetryWrapper`, since we can't add this logic to the runner, the cleanest way I thought of for doing this work. Maybe not the best name, would like suggestions for that.